Web Docs - Stepper Component (#2727)
hashicorp/design-system: website/docs/components/stepper/nav/partials/code/how-to-use.md

<Hds::Stepper::Nav @currentStep={{1}} @ariaLabel="Status" as |S|>
  <S.Step>
    <:title>Step 1</:title>
  </S.Step>
  <S.Step>
    <:title>Step 2</:title>
  </S.Step>
  <S.Step>
    <:title>Step 3</:title>
  </S.Step>

  <S.Panel>Content 1</S.Panel>
  <S.Panel>Content 2</S.Panel>
  <S.Panel>Content 3</S.Panel>
</Hds::Stepper::Nav>
```

### Interactivity

The `@isInteractive` argument is used to control if navigation between steps is allowed using the steps. If `true`, users can navigate to completed steps by clicking on the Step. Users can only navigate to completed steps, not incomplete ones.

When a click on a step occurs, the `@onStepChange` handler can be used to pass a custom function to update the current step number.

```handlebars
<Hds::Stepper::Nav
  @currentStep={{this.demoCurrentStep}}
  @isInteractive={{true}}
  @ariaLabel="Interactive"
  @onStepChange={{this.demoOnStepChange}}
  as |S|
>
  <S.Step>
    <:title>Step 1</:title>
  </S.Step>
  <S.Step>
    <:title>Step 2</:title>
  </S.Step>
  <S.Step>
    <:title>Step 3</:title>
  </S.Step>

  <S.Panel>Content 1</S.Panel>
  <S.Panel>Content 2</S.Panel>
  <S.Panel>Content 3</S.Panel>
</Hds::Stepper::Nav>
```

> ### Without panels

Reviewer: @shleewhite I added a section for using the nav without panels, which we've discussed earlier. Please let me know if the accessibility notes here look correct, or if any changes need to be made to the accessibility file as well.

Reviewer 2: Looks good!

When `@isInteractive` is set to `false`, the component steps do not allow navigation between steps.

```handlebars
<Hds::Stepper::Nav
  @currentStep={{1}}
  @isInteractive={{false}}
  @ariaLabel="Interactive"
  as |S|
>
  <S.Step>
    <:title>Step 1</:title>
  </S.Step>
  <S.Step>
    <:title>Step 2</:title>
  </S.Step>
  <S.Step>
    <:title>Step 3</:title>
  </S.Step>

  <S.Panel>Content 1</S.Panel>
  <S.Panel>Content 2</S.Panel>
  <S.Panel>Content 3</S.Panel>
</Hds::Stepper::Nav>
```

### Without panels

The component can be used without the `[S].Panel` contextual component, or `<:body>` named block. In this use case, the component operates similar to a static list. This can be used if the content of the steps is outside of the component.

!!! Info

Due to accessibility implications, `@isInteractive` must be set to `false` when using the component without panels.

!!!

```handlebars
<Hds::Stepper::Nav
  @isInteractive={{false}}
  @ariaLabel="Standalone contextual components"
  as |S|
>
  <S.Step>
    <:title>Step 1</:title>
  </S.Step>
  <S.Step>
    <:title>Step 2</:title>
  </S.Step>
  <S.Step>
    <:title>Step 3</:title>
  </S.Step>
</Hds::Stepper::Nav>
Steps content
```

> ```handlebars

Reviewer 2: Issue: is there supposed to be 2 code examples right next to each other here?
<img width="441" alt="Screenshot 2025-03-13 at 3 04 15 PM" src="https://github.com/user-attachments/assets/320f9e22-c9ab-4627-bd04-6ae76dea5a8e" />

Reviewer: I was giving an example for both the contextual component and steps arg use cases, but agree it's unnecessary. Removed the last example.

## How to use this component

The Stepper Nav component is used to track the linear progress of a user in a given on-page flow.

> To use this component, you must either use the contextual components for Step and Panel, or use the `@steps` argument.

Reviewer 3: ```suggestion
To use this component, you can either pass in the contextual Step and Panel components, or use the `@steps` argument to pass in an array of steps.
```

### Using contextual components

The `Hds::Stepper::Nav::Step` and `Hds::Stepper::Nav::Panel` are yielded as contextual components. The number of steps and panels added must be equal.

```handlebars
<Hds::Stepper::Nav as |S|>
  <S.Step>
    <:title>Step 1</:title>
  </S.Step>
  <S.Step>
    <:title>Step 2</:title>
  </S.Step>
  <S.Step>
    <:title>Step 3</:title>
  </S.Step>

  <S.Panel>Content 1</S.Panel>
  <S.Panel>Content 2</S.Panel>
  <S.Panel>Content 3</S.Panel>
</Hds::Stepper::Nav>
```

### Using the @steps argument

> When using the `@steps` argument, use the named block `<:body>` to pass in step content.

Reviewer 2: Suggestion: could you elaborate on a use case where it would be better to use `@steps` or the contextual components?

Reviewer: Elaborated on this statement, let me know if it needs altering.

```handlebars
<Hds::Stepper::Nav
  @steps={{array
    (hash title="Step 1")
    (hash title="Step 2")
    (hash title="Step 3")
  }}
>

>   <:body>Content</:body>

Reviewer 3: @dchyun Perhaps you could include this example from your component PR in the `:body` here?

```
  <:body>
    {{#if (eq this.currentStep 0)}}
      Step 1 content
    {{else}}
      Step 2 content
    {{/if}}
  </:body>
```

</Hds::Stepper::Nav>
```

### Step state

> The state of steps is controlled automatically through the `@currentStep` argument. The argument is base-indexed and sets the active step to the value provided. The status of a step is determined in the following ways in relation to the `@currentStep`.

Reviewer 2: ```suggestion
The state of steps is controlled automatically through the `@currentStep` argument. The argument is zero-indexed and sets the active step to the value provided. The status of a step is determined in the following ways in relation to the `@currentStep`.
```

- A step equal to `@currentStep` is active
- All steps less than `@currentStep` are completed
- All steps greater than `@currentStep` are incomplete

> By default `@currentStep` is equal to 0.

Reviewer 3: ```suggestion
By default, `@currentStep` is equal to 0.
```

```handlebars
<Hds::Stepper::Nav @currentStep={{1}} as |S|>
  <S.Step>
    <:title>Step 1</:title>
  </S.Step>
  <S.Step>
    <:title>Step 2</:title>
  </S.Step>
  <S.Step>
    <:title>Step 3</:title>
  </S.Step>

  <S.Panel>Content 1</S.Panel>
  <S.Panel>Content 2</S.Panel>
  <S.Panel>Content 3</S.Panel>
</Hds::Stepper::Nav>
```

### Interactivity

> The `@isInteractive` argument is used to control if navigation between steps is allowed using the steps. If `true`, users can navigate to completed steps by clicking on the Step. Users can only navigate to completed steps, not incomplete 1s.

Reviewer: Thanks for catching my find and replace mistake 😅

When a click on a step occurs, the `@onStepChange` handler can be used to pass a custom function to update the current step number.

```handlebars
<Hds::Stepper::Nav
  @currentStep={{1}}
  @isInteractive={{true}}

>   @onStepChange={{this.updateStepNumber}}

Reviewer 3: @dchyun Perhaps you could write a really simple script called `updateStepNumber` for this example to add some really basic interactivity? (You don't need to include the code within then text of the documentation. It would just be used to do something when consumers click on the current rendered code example.)

Reviewer: Added interactivity to this example for `@onStepChange`, which allows interaction through the steps.
